### PR TITLE
http/2: correctly handle 100-continue

### PR DESCRIPTION
**Pull request by mattklein123**
Previously we did not correctly handle "expect: 100-continue" in the http/2
codec. Because of protocol differences, we handle 100-continue directly in the
codec. This change implements support in the http/2 codec similar to http/1.1
where we immediately send a continue response.

This commit also adds partial support for the client codec for 1xx responses.
The code will raise decodeHeaders(...) twice in the 1xx response case. This
is good enough for testing, but will probably crash higher layers if a client
actually responded this way. Right now we effectively trust clients so this
can be handled in a follow up if necessary.

fixes https://github.com/lyft/envoy/issues/561

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team @jpinner-lyft @andrewwebber

File: source/common/http/http2/codec_impl.cc [hunk 1]
```diff
@@ -283,17 +291,64 @@ int ConnectionImpl::onFrameReceived(const nghttp2_frame* frame) {
...
-        stream->decoder_->decodeTrailers(std::move(stream->headers_));
+        if (!stream->waiting_for_final_headers_) {
+          ASSERT(stream->remote_end_stream_);
+          stream->decoder_->decodeTrailers(std::move(stream->headers_));
+        } else {
+          ASSERT(!nghttp2_session_check_server_session(session_));
+          stream->waiting_for_final_headers_ = false;
+
+          // This can only happen in the client case in a response, when we received a 1xx to
+          // start out with. In this case, raise as headers. nghttp2 message checking guarentees
```
Comment (ccaraman) [inline source/common/http/http2/codec_impl.cc]: typo guarantees

**Comment by mattklein123** [comment]
@htuch could you give this a look also please.

File: test/common/http/http2/codec_impl_test.cc [hunk 2]
```diff
@@ -62,6 +62,72 @@ class Http2CodecImplTest : public testing::TestWithParam<uint64_t> {
...
+  EXPECT_CALL(request_decoder, decodeData(_, true));
+  Buffer::OwnedImpl hello("hello");
+  request_encoder.encodeData(hello, true);
+
+  TestHeaderMapImpl response_headers{{":status", "200"}};
+  EXPECT_CALL(response_decoder, decodeHeaders_(HeaderMapEqual(&response_headers), true));
+  response_encoder->encodeHeaders(response_headers, true);
+}
+
+TEST_P(Http2CodecImplTest, ExpectContinueTrailersResponse) {
```
Comment (htuch) [inline test/common/http/http2/codec_impl_test.cc]: Do you cover the case where the non-100 :status is in the HEADERS immediately following the 100 :status?

File: source/common/http/filter/ratelimit.cc [hunk 3]
```diff
@@ -11,8 +11,9 @@
 namespace Http {
 namespace RateLimit {
 
-const Http::HeaderMapPtr Filter::TOO_MANY_REQUESTS_HEADER{new Http::HeaderMapImpl{
-    {Http::Headers::get().Status, std::to_string(enumToInt(Code::TooManyRequests))}}};
+const std::unique_ptr<const Http::HeaderMap> Filter::TOO_MANY_REQUESTS_HEADER{
+    new Http::HeaderMapImpl{
```
Comment (mattklein123) [inline source/common/http/filter/ratelimit.cc]: We do correctly join all threads before exiting main. I'm not a huge fan of the raw pointer approach because AFAIK this will cause the heap leak detector to fail on standard settings. I don't feel super strongly about this, but we do this in other places also. Thoughts?

File: source/common/http/http2/codec_impl.cc [hunk 4]
```diff
@@ -283,17 +291,64 @@ int ConnectionImpl::onFrameReceived(const nghttp2_frame* frame) {
...
+    if (frame->headers.cat == NGHTTP2_HCAT_REQUEST && stream->headers_->Expect() &&
+        0 == StringUtil::caseInsensitiveCompare(stream->headers_->Expect()->value().c_str(),
+                                                Headers::get().ExpectValues._100Continue.c_str())) {
+      // Deal with expect: 100-continue here since higher layers are never going to do anything
+      // other than say to continue so that we can respond before request complete if necessary.
+      std::vector<nghttp2_nv> final_headers;
+      StreamImpl::buildHeaders(final_headers, *CONTINUE_HEADER);
+      int rc = nghttp2_submit_headers(session_, 0, stream->stream_id_, nullptr, &final_headers[0],
+                                      final_headers.size(), nullptr);
+      ASSERT(rc == 0);
```
Comment (mattklein123) [inline source/common/http/http2/codec_impl.cc]: Based on logic flow, this function should never fail, so I made it a debug only assert. We do this in a few other places in this file.

File: test/common/http/http2/codec_impl_test.cc (hunk 2 above)
Comment (mattklein123) [inline test/common/http/http2/codec_impl_test.cc]: What do you mean exactly? Like if Envoy immediately responds with 403 right after sending 100?

**Comment by mattklein123** [comment]
> Do any Envoy use cases motivate forwarding the 100-continue to the backend? It seems that in general the original intent of the expect header was to have the proxy forward it and then determine from the backend whether or not to reply 100.

> This might be legacy by now, and it certainly would hurt performance to round-trip unconditionally.

I don't know of any currently, and we already do direct response for HTTP/1.1. Being able to correctly proxy the 1xx through from server to client would be a giant change. My preference would be to do this and then if we actually find a case that requires 100 continue server support (I really doubt this will happen) we can cover it then.

File: source/common/http/http2/codec_impl.cc [hunk 5]
```diff
@@ -283,17 +291,64 @@ int ConnectionImpl::onFrameReceived(const nghttp2_frame* frame) {
...
+        } else {
+          ASSERT(!nghttp2_session_check_server_session(session_));
+          stream->waiting_for_final_headers_ = false;
+
+          // This can only happen in the client case in a response, when we received a 1xx to
+          // start out with. In this case, raise as headers. nghttp2 message checking guarantees
+          // proper flow here.
+          // TODO(mattklein123): Higher layers don't currently deal with a double decodeHeaders()
+          // call and will probably crash. We do this for testing the server case. We correctly
+          // handle the server case. In the future, if needed, we can properly handle 1xx in higher
```
Comment (htuch) [inline source/common/http/http2/codec_impl.cc]: Not clear what is meant by the server case here - how does the server receive a 100 response, isn't it the one generating the 100 response?

File: test/common/http/http2/codec_impl_test.cc [hunk 6]
```diff
@@ -62,6 +62,72 @@ class Http2CodecImplTest : public testing::TestWithParam<uint64_t> {
   ConnectionWrapper server_wrapper_;
 };
 
+TEST_P(Http2CodecImplTest, ExpectContinueHeadersOnlyResponse) {
+  MockStreamDecoder response_decoder;
+  StreamEncoder& request_encoder = client_.newStream(response_decoder);
+
+  MockStreamDecoder request_decoder;
+  StreamEncoder* response_encoder;
+  EXPECT_CALL(server_callbacks_, newStream(_))
```
Comment (htuch) [inline test/common/http/http2/codec_impl_test.cc]: Suggest refactoring this encoder/decoder setup into the Http2CodecImplTest class, it seems the boiler plate is now in quite a few test cases.

File: test/common/http/http2/codec_impl_test.cc (hunk 2 above)
Comment (htuch) [inline test/common/http/http2/codec_impl_test.cc]: Yeah, but thinking about it a bit more, I think you have good coverage of the new code paths in these tests, so it's good as is.

File: test/common/upstream/outlier_detection_impl_test.cc [hunk 7]
```diff
@@ -77,6 +77,7 @@ TEST_F(OutlierDetectorImplTest, DetectorStaticConfig) {
   )EOF";
 
   Json::ObjectPtr custom_config = Json::Factory::LoadFromString(json);
+  EXPECT_CALL(*interval_timer_, enableTimer(std::chrono::milliseconds(100)));
```
Comment (htuch) [inline test/common/upstream/outlier_detection_impl_test.cc]: Is this a related change?

File: source/common/http/filter/ratelimit.cc (hunk 3 above)
Comment (htuch) [inline source/common/http/filter/ratelimit.cc]: There's also the concern about non-deterministic ordering of initialization. Not really a concern for simple local objects like this. 

I think that when done sanely (e.g. don't initialize static data that depends on other static data), and given we have proper thread joining on exit, it's probably fine.

It would make sense to document this in https://github.com/lyft/envoy/blob/master/CONTRIBUTING.md though.

Comment (mattklein123) [inline source/common/http/filter/ratelimit.cc]: Yes, for init, we need to more careful, and we do handle more complex cases in init by having functions that return status variables. I will do a doc update.

File: test/common/upstream/outlier_detection_impl_test.cc (hunk 7 above)
Comment (mattklein123) [inline test/common/upstream/outlier_detection_impl_test.cc]: Just a missing EXPECT_CALL from a previous change. I added it here since innocuous. Can move to a separate PR if you prefer.

File: source/common/http/http2/codec_impl.cc (hunk 4 above)
Comment (mattklein123) [inline source/common/http/http2/codec_impl.cc]: None of the Envoy code is safe for memory allocation failure, so we don't handle similar errors anywhere and assume we will shortly crash if OOM. (This is a design philosophy, we should discuss this outside of the context of this change if there are concerns).

Comment (htuch) [inline source/common/http/http2/codec_impl.cc]: Sure, we can discuss this elsewhere (I do worry about libraries using ENOMEM in non-OOM scenarios, as opposed to tcmalloc). In any case, could you also include this design philosophy in https://github.com/lyft/envoy/blob/master/CONTRIBUTING.md when updating? Thanks.

Comment (mattklein123) [inline source/common/http/http2/codec_impl.cc]: Yes that is a valid concern and might necessitate some paranoia checking in certain cases which would crash us if we receive ENOMEM, etc. I will open a GH issue to track as well as updating contributing guide.

File: test/common/upstream/outlier_detection_impl_test.cc (hunk 7 above)
Comment (htuch) [inline test/common/upstream/outlier_detection_impl_test.cc]: Sure, sneak it in, just curious what was happening.

**Comment by mattklein123** [comment]
@lyft/network-team @htuch updated
